### PR TITLE
Add openstack regression test runner

### PR DESCRIPTION
**Pull request by dosaboy**
Run the Openstack release tests from charmed-openstack-tests.

### CODE REVIEW

**Comment by dosaboy** [comment]
Currently getting the following error because ceph units do not have enough storage:

```
root@juju-efd3bb-zaza-c4f5cb82f94b-4:/home/ubuntu# ceph -s
  cluster:
    id:     d6be62ca-7cde-11ef-bc82-5d4897b1b779
    health: HEALTH_ERR
            clock skew detected on mon.juju-efd3bb-zaza-c4f5cb82f94b-6
            3 full osd(s)
            7 pool(s) full
 
  services:
    mon: 3 daemons, quorum juju-efd3bb-zaza-c4f5cb82f94b-4,juju-efd3bb-zaza-c4f5cb82f94b-5,juju-efd3bb-zaza-c4f5cb82f94b-6 (age 91m)
    mgr: juju-efd3bb-zaza-c4f5cb82f94b-6(active, since 90m), standbys: juju-efd3bb-zaza-c4f5cb82f94b-4, juju-efd3bb-zaza-c4f5cb82f94b-5
    mds: 1/1 daemons up
    osd: 3 osds: 3 up (since 89m), 3 in (since 90m)
 
  data:
    volumes: 1/1 healthy
    pools:   7 pools, 157 pgs
    objects: 2.98k objects, 23 GiB
    usage:   69 GiB used, 3.4 GiB / 72 GiB avail
    pgs:     157 active+clean
```

**Comment by dosaboy** [comment]
they are getting staging-cpu1-ram4-disk20 flavor atm

**Comment by dosaboy** [comment]
bundle has:

```
  ceph-osd:
    charm: ch:ceph-osd
    num_units: 3
    options:
      source: *source
    storage:
      osd-devices: cinder,24G
    constraints: mem=4096
    channel: *ceph-channel
```

**Comment by dosaboy** [comment]
```
# ceph df
--- RAW STORAGE ---
CLASS    SIZE    AVAIL    USED  RAW USED  %RAW USED
hdd    72 GiB  3.4 GiB  69 GiB    69 GiB      95.34
TOTAL  72 GiB  3.4 GiB  69 GiB    69 GiB      95.34
 
--- POOLS ---
POOL              ID  PGS   STORED  OBJECTS     USED   %USED  MAX AVAIL
.mgr               1    1  449 KiB        2  1.3 MiB  100.00        0 B
ceph-fs_data       2    4      0 B        0      0 B       0        0 B
ceph-fs_metadata   3   16  4.0 KiB       22   96 KiB  100.00        0 B
manila-ganesha     4   32      4 B        3   12 KiB  100.00        0 B
gnocchi            5    8     14 B        1   12 KiB  100.00        0 B
glance             6   64   23 GiB    2.95k   68 GiB  100.00        0 B
cinder-ceph        7   32      0 B        0      0 B       0        0 B
```

**Comment by dosaboy** [comment]
names of images synced to glance:
```
+---------------------------------------------------------------+
| name                                                          |
+---------------------------------------------------------------+
| cirros_alt                                                    |
| bionic                                                        |
| fedora-coreos                                                 |
| auto-sync/ubuntu-jammy-22.04-amd64-server-20240912-disk1.img  |
| amphora-haproxy-x86_64-ubuntu-22.04-20240912                  |
| noble                                                         |
| fedora-coreos                                                 |
| cirros                                                        |
| jammy                                                         |
| auto-sync/ubuntu-focal-20.04-amd64-server-20240821-disk1.img  |
| auto-sync/ubuntu-xenial-16.04-amd64-server-20211001-disk1.img |
| auto-sync/ubuntu-trusty-14.04-amd64-server-20191107-disk1.img |
| focal                                                         |
| auto-sync/ubuntu-bionic-18.04-amd64-server-20230607-disk1.img |
+---------------------------------------------------------------+
```

File: openstack/tools/openstack_regression_tests_runner.sh [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+done
+
+if [[ -z $FUNC_TEST_TARGET ]]; then
+   echo "ERROR: must provide a target name with --func-test-target"
+   exit 1
+fi
+
+# This is required for magnum tests and zaza will look in swift if it is not cached so we need to cache it first.
+mkdir -p $IMAGES_PATH
+if ! [[ -e $IMAGES_PATH/fedora-coreos-35.qcow2 ]]; then
```
Comment (pponnuvel) [inline openstack/tools/openstack_regression_tests_runner.sh:59]: `-f` is better here. Also can move the "not" inside:
```
if [[ ! -f $IMAGES_PATH/fedora-coreos-35.qcow2 ]]
```

File: openstack/tools/openstack_regression_tests_runner.sh [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+    cd zosci-config
+    git checkout master
+    git pull
+else
+    git clone https://github.com/openstack-charmers/zosci-config
+fi
+)
+
+TOOLS_PATH=$(realpath $(dirname $0))/func_test_tools
+CHARM_PATH=$(pwd)
```
Comment (pponnuvel) [inline openstack/tools/openstack_regression_tests_runner.sh:80]: `CHARM_PATH=$PWD` would be better.

File: openstack/tools/openstack_regression_tests_runner.sh [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+
+echo "Running regression tests"
+
+source ~/novarc
+export {,TEST_}CIDR_EXT=`openstack subnet show subnet_${OS_USERNAME}-psd-extra -c cidr -f value`
```
Comment (pponnuvel) [inline openstack/tools/openstack_regression_tests_runner.sh:85]: Backticks are used in some places while `$(..)` are used in others. `$(..)` is generally recommended, so can stick to that for consistency as well.

See https://mywiki.wooledge.org/BashFAQ/082

File: openstack/tools/openstack_regression_tests_runner.sh [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+
+LOGFILE=$(mktemp --suffix=-charm-func-test-results)
+(
+    # Ensure charmed-openstack-tester checked out and up-to-date
+    if [[ -d $HOME/charmed-openstack-tester ]]; then
+        cd $HOME/charmed-openstack-tester
+        git checkout master
+        git pull
+    else
+        cd $HOME
```
Comment (pponnuvel) [inline openstack/tools/openstack_regression_tests_runner.sh:127]: Why not just `cd`? Just asking as you did that [here](https://github.com/canonical/stsstack-bundles/pull/254/files#diff-df917a4f34332b87c75ba0996575e9c7bec5dce969017f0dfdcf4b20bb9dd0a8R69) ;-)

File: openstack/tools/openstack_regression_tests_runner.sh [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+    fi
+
+    # If a func test pr is provided switch to that pr.
+    if [[ -n $FUNC_TEST_PR ]]; then
+        # We use the zosci-config tools to do this.
+        MSG=$(echo "Func-Test-Pr: https://github.com/openstack-charmers/zaza-openstack-tests/pull/$FUNC_TEST_PR"| base64)
+        ~/zosci-config/roles/handle-func-test-pr/files/process_func_test_pr.py -f ./test-requirements.txt "$MSG"
+    fi
+
+    fail=false
```
Comment (pponnuvel) [inline openstack/tools/openstack_regression_tests_runner.sh:139]: What's this `fail` for?

**Review by pponnuvel — APPROVED** [review]
Some nit picks. Otherwise looks good to me.

File: openstack/tools/openstack_regression_tests_runner.sh (hunk 1 above)
Comment (dosaboy) [reply openstack/tools/openstack_regression_tests_runner.sh:59]: done.

File: openstack/tools/openstack_regression_tests_runner.sh (hunk 2 above)
Comment (dosaboy) [reply openstack/tools/openstack_regression_tests_runner.sh:80]: done.

File: openstack/tools/openstack_regression_tests_runner.sh (hunk 3 above)
Comment (dosaboy) [reply openstack/tools/openstack_regression_tests_runner.sh:85]: done. also did the same in charmed_openstack_functest_runner.sh

File: openstack/tools/openstack_regression_tests_runner.sh (hunk 4 above)
Comment (dosaboy) [reply openstack/tools/openstack_regression_tests_runner.sh:127]: ack

File: openstack/tools/openstack_regression_tests_runner.sh (hunk 5 above)
Comment (dosaboy) [reply openstack/tools/openstack_regression_tests_runner.sh:139]: hangover from copy+paste have cleaned up

**Pull request merged**